### PR TITLE
lite block fixes

### PR DESCRIPTION
**Pull request by michael-herwig**
This is how far I got until a friend shown up that wanna drink with me :). Please have a look and give some feedback.

This should fix all lite block issues except the alternative chain problem as the missing transactions notification still pushes only into the transaction pool and some transactions may still are missing that would be valid on alternative chains.

### CODE REVIEW

File: src/CryptoNoteCore/Core.cpp [hunk 1]
```diff
@@ -863,6 +863,18 @@ std::string Core::getPaymentIDFromExtra(const std::vector<uint8_t> &extra)
     return std::string();
 }
 
+boost::optional<BinaryArray> Core::getTransaction(const Crypto::Hash& hash) const {
```
[the remark this answers is not in the document]
Comment (michael-herwig) [reply src/CryptoNoteCore/Core.cpp]: does cmake enforce c++17?

[the remark this answers is not in the document]
Comment (michael-herwig) [reply src/CryptoNoteCore/Core.cpp]: ok it does gonna change that

**Comment by brandonlehmann** [comment]
For clarity, this should resolve #727 and #728 correct?

**Comment by michael-herwig** [comment]
yes and partially #729, will add some explanation tomorrow

**Comment by brandonlehmann** [comment]
Awesome. Good work!

**Comment by michael-herwig** [comment]
I think this should do it for the three tickets. I didn't test it yet but the implementation should do the following things now:

* missing transactions of a lite block will only be requested from the publisher
* if a missing transaction request is sent the initial lite block receiver caches the lite block request as well as the request of the missing transaction
* if a node gets a new transactions it checks if there is a cached pending lite block request for this peer, if so the transactions are not considered as new pool transaction. Instead, they are pushed into the lite block routine.
* the lite block routine now can handle an additional context of provided transactions, it will search for provided transactions, blockchain transaction and pool transaction to fill all necessary transaction details.
* the push lite block routine has some additional sanity checks for checking if all requested missing transactions were provided or a peer tries to push a new lite block while one is already pending

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 2]
```diff
@@ -413,18 +413,22 @@ int CryptoNoteProtocolHandler::handle_notify_new_transactions(int command, NOTIF
...
-  }
-
-  if (arg.txs.size()) {
-    //TODO: add announce usage here
-    relay_post_notify<NOTIFY_NEW_TRANSACTIONS>(*m_p2p, arg, &context.m_connection_id);
+  if(context.m_pending_lite_block.has_value()) {
+      logger(Logging::TRACE) << context << " Pending lite block detected, handling request as missing lite block transactions response";
+      return doPushLiteBlock(context.m_pending_lite_block->request, context, std::move(arg.txs));
+  } else {
+      for (auto tx_blob_it = arg.txs.begin(); tx_blob_it != arg.txs.end();) {
```
Comment (zpalmtree) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: Might be less error prone and a tad easier to read if we use `std::remove_if` here?

E.g.

```cpp
const auto it = std::remove_if(arg.txs.begin(), arg.txs.end(), [&m_core](const auto &tx)
{
    bool failed = !m_core.addTransactionToPool(tx);

    if (failed)
    {
        logger(Logging::DEBUGGING) << context << "Tx verification failed";
    }

    return failed;
});

if (it != arg.txs.end())
{
    arg.txs.erase(it, arg.txs.end());
}
```

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 3]
```diff
@@ -413,18 +413,22 @@ int CryptoNoteProtocolHandler::handle_notify_new_transactions(int command, NOTIF
...
+          if (!m_core.addTransactionToPool(*tx_blob_it)) {
+              logger(Logging::DEBUGGING) << context << "Tx verification failed";
+              tx_blob_it = arg.txs.erase(tx_blob_it);
+          } else {
+              ++tx_blob_it;
+          }
+      }
+      if (arg.txs.size()) {
```
Comment (zpalmtree) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: I'd prefer explicitly checking the size here.

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 4]
```diff
@@ -567,6 +571,121 @@ int CryptoNoteProtocolHandler::processObjects(CryptoNoteConnectionContext& conte
...
+    BlockTemplate newBlockTemplate;
+    if(!fromBinaryArray(newBlockTemplate, arg.blockTemplate)) { // deserialize blockTemplate
+        logger(Logging::WARNING) << context << "Deserialization of Block Template failed, dropping connection" ;
+        context.m_state = CryptoNoteConnectionContext::state_shutdown;
+        return 1;
+    }
+
+    std::unordered_map<Crypto::Hash, BinaryArray> provided_txs;
+    provided_txs.reserve(missingTxs.size());
+    for(std::size_t i = 0; i < missingTxs.size(); ++i) {
```
Comment (zpalmtree) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: Is there any reason to use a for loop over a ranged based loop here?

E.g.

```cpp
for (const auto &tx: missingTxs)
{
    provided_txs[getBinaryArrayHash(tx)] = tx;
}
```

Could use std::move here, but personally I'm not a fan of std::move when we're using the tx on the same line - makes it easy to invoke undefined behaviour if we refactor slightly.

I'm not seeing us using the `CachedTransaction i_provided_transaction{std::move(missingTxs[i])};` line either?

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 5]
```diff
@@ -567,6 +571,121 @@ int CryptoNoteProtocolHandler::processObjects(CryptoNoteConnectionContext& conte
...
+    }
+
+    std::vector<BinaryArray> have_txs;
+    std::vector<Crypto::Hash> need_txs;
+
+    if(context.m_pending_lite_block.has_value()) {
+        for(const auto& requestedTxHash : context.m_pending_lite_block->missed_transactions) {
+            if(provided_txs.find(requestedTxHash) == provided_txs.end()) {
+                logger(Logging::DEBUGGING) << context << "Peer didn't provide a missing transaction, previously "
+                                                         "acquired for a lite block, dropping conneciton.";
```
Comment (zpalmtree) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: Typo, `conneciton`.

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 6]
```diff
@@ -567,6 +571,121 @@ int CryptoNoteProtocolHandler::processObjects(CryptoNoteConnectionContext& conte
...
+
+    /*
+   * here we are finding out which txs are
+   * present in the pool and which are not
+   * further we check for transactions in
+   * the blockchain to accept alternative
+   * blocks.
+   */
+    for (const auto transactionHash: newBlockTemplate.transactionHashes) {
+        auto providedSerach = provided_txs.find(transactionHash);
```
Comment (zpalmtree) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: Typo, `providedSerach`.

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp (hunk 4 above)
Comment (michael-herwig) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: Yes I forgot to delete the CachedTransaction line there. For using a range based loop, if you want to move the object I tend to use a classic for in favor to not use `for(auto&& ...)`. Anway the double move is actually a bug in there gotta fix that.

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp (hunk 6 above)
Comment (michael-herwig) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: `const auto` is missing `&` as well

**Comment by zpalmtree** [comment]
Looks good. Thanks for the PR! Do we want to spin up a testnet @brandonlehmann @RocksteadyTC ?

**Comment by brandonlehmann** [comment]
Probably a good idea.

**Comment by RocksteadyTC** [comment]
I'm free for testnet today if we want to give this a shot.

**Pull request merged**